keps/sig-instrumentation/2845-deprecate-klog-specific-flags-in-k8s-components/README.md

```diff
@@ -0,0 +1,239 @@
+was forked to klog and multiple improvements were implemented, but features put
+into klog only piled up and were never removed. Introduction of alternative log
+formats like JSON created a conundrum, should we implement all klog features for
+JSON? Most of them don't make sense and method for their configuration leaves
+much to be desired. Klog features are controlled by set of global flags that
+remain last bastion of global state in k/k repository. Those flags don't have a
+single naming standard (some start with log prefix, some not), don't comply to
+k8s flag naming (use underscore instead of hyphen) and many other problems. We
+need to revisit how logging configuration is done in klog so it can work with
+alternative log formats and comply with current best practices.
+
+Large number of features added to klog has lead to large drop in quality. 
+[#90804](https://github.com/kubernetes/kubernetes/pull/90804) shows example 
+where kubeup (canonical way to deploy kubernetes for testing) could not use 
+klog feature to write log files due to scalability issues. The maintainers of 
+klog decided it's easier to re-implementing a canonical klog feature in external
+project than debugging the underlying problem.
+
+As for what configuration options should be standardized for all logging formats
+I would look into 12 factor app standard (https://12factor.net/). It defines
+logs as steams of events and discourages applications from taking on
+responsibility for log file management, log rotation and any other processing
+that can be done externally. This is something that Kubernetes already
+encourages by collecting stdout and stderr logs and making them available via
+kubectl logs. It's somewhat confusing that K8s components don't comply to K8s
+best practices.
+
+### Goals
+
+* Unblock development of alternative logging formats
+* Improve quality of logging in k8s core components
+* Logging should use configuration mechanism developed by WG component-standard
+
+### Non-Goals
+
+* Change klog output format
+
+## Proposal
+
+I propose that Kubernetes core components (kube-apiserver, kube-scheduler,
+kube-controller-manager, kubelet) should drop flags that extend
+logging over events streams or klog specific features. This change should be 
+scoped to only those components and not affect broader klog community.
+
+Flags that should be deprecated:
+
+* --log-dir, --log-file, --log-flush-frequency - responsible for writing to 
+  files and syncs to disk. 
+  Motivation: Remove complexity to make alternative loggers easier to implement
+  and reducing feature surface to improve quality.
+* --logtostderr, --alsologtostderr, --one-output, --stderrthreshold - 
+  responsible enabling/disabling writing to stderr (vs file). 
+  Motivation: Not needed if writing to files is removed.
+* --log-file-max-size, --skip-log-headers - responsible configuration of file 
+  rotation. 
+  Motivation: Not needed if writing to files is removed.
```

Comment:
We should probably add somewhere a very quick detail of why the ability to write to files is not critical (shell redirection or systemd service management which most people are already using, e.g.)

Reply:
ack